File: src/WearablePayments.App/Services/ApiService.cs
namespace WearablePayments.App.Services;

using System.Net.Http.Headers;
using System.Net.Http.Json;

public class ApiService
{
    private readonly HttpClient _http;
    private string? _token;

    public ApiService(HttpClient http) => _http = http;

    public void SetToken(string token)
    {
        _token = token;
        _http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
    }

    public Task<T?> GetAsync<T>(string path) => _http.GetFromJsonAsync<T>(path);

    public async Task<T?> PostAsync<T>(string path, object body)
    {
        var response = await _http.PostAsJsonAsync(path, body);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadFromJsonAsync<T>();
    }

    public async Task DeleteAsync(string path)
    {
        var response = await _http.DeleteAsync(path);
        response.EnsureSuccessStatusCode();
    }
}

File: src/WearablePayments.Flutter/lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class ApiService {
  static const _base = 'http://localhost:5100';
  String _token = '';

  void setToken(String token) => _token = token;
  String get token => _token;

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        if (_token.isNotEmpty) 'Authorization': 'Bearer $_token',
      };

  Future<Map<String, dynamic>> get(String path) async {
    final res = await http.get(Uri.parse('$_base/$path'), headers: _headers);
    _check(res);
    return jsonDecode(res.body);
  }

  Future<List<dynamic>> getList(String path) async {
    final res = await http.get(Uri.parse('$_base/$path'), headers: _headers);
    _check(res);
    return jsonDecode(res.body);
  }

  Future<Map<String, dynamic>> post(String path, Map<String, dynamic> body) async {
    final res = await http.post(
      Uri.parse('$_base/$path'),
      headers: _headers,
      body: jsonEncode(body),
    );
    _check(res);
    return jsonDecode(res.body);
  }

  Future<void> delete(String path) async {
    final res = await http.delete(Uri.parse('$_base/$path'), headers: _headers);
    _check(res);
  }

  void _check(http.Response res) {
    if (res.statusCode >= 400) {
      final body = jsonDecode(res.body);
      throw Exception(body['message'] ?? 'Request failed (${res.statusCode})');
    }
  }
}
